### PR TITLE
feat(react-charts): Stacked HBC And Negative X value support

### PR DESCRIPTION
**Pull request by Anush2303**
## New Behavior

Stacked HBC With axis is implemented. Support for negative x values added.
Corresponding v8 commits:
https://github.com/microsoft/fluentui/commit/f5d76970187c11353c3387e2074949f376cea9f8
https://github.com/microsoft/fluentui/commit/5c401aa875ddb230d073a0091a16481d0ae58ad4

## Related Issue(s)

- Fixes #

### CODE REVIEW

**Comment by AtishayMsft** [comment]
Can you share playwright test results

File: packages/charts/react-charts/library/src/components/HorizontalBarChartWithAxis/HorizontalBarChartWithAxis.tsx [hunk 1]
```diff
@@ -598,6 +724,33 @@ export const HorizontalBarChartWithAxis: React.FunctionComponent<HorizontalBarCh
...
+    chartType: ChartTypes,
+    isRTL: boolean,
+    xAxisType: XAxisTypes,
+    barWidth: number,
+    tickValues: Date[] | number[] | undefined,
+    shiftX: number,
+  ) {
+    let domainNRangeValue: IDomainNRange;
+    if (xAxisType === XAxisTypes.NumericAxis) {
+      domainNRangeValue = domainRangeOfNumericForHorizontalBarChartWithAxis(
```
Comment (AtishayMsft) [inline packages/charts/react-charts/library/src/components/HorizontalBarChartWithAxis/HorizontalBarChartWithAxis.tsx:740]: Does this function need to be in utilities now.

Comment (Anush2303) [reply packages/charts/react-charts/library/src/components/HorizontalBarChartWithAxis/HorizontalBarChartWithAxis.tsx:740]: Yes, we can shift this also to base file but then we will have to shift computeLongestBars function also which is used by it. It is better if we invoke this from utilities file

Comment (Anush2303) [reply packages/charts/react-charts/library/src/components/HorizontalBarChartWithAxis/HorizontalBarChartWithAxis.tsx:740]: We can remove thus case from switch case.

File: packages/charts/react-charts/library/src/components/HorizontalBarChartWithAxis/HorizontalBarChartWithAxis.tsx [hunk 2]
```diff
@@ -167,10 +174,40 @@ export const HorizontalBarChartWithAxis: React.FunctionComponent<HorizontalBarCh
     xElement?: SVGElement | null,
     yElement?: SVGElement | null,
   ) {
-    return (_bars =
+    const stackedChartData = groupChartDataByYValue(_points);
```
Comment (srmukher) [inline packages/charts/react-charts/library/src/components/HorizontalBarChartWithAxis/HorizontalBarChartWithAxis.tsx:177]: Expensive calculations can be memoized 
For example:
const groupedData = React.useMemo(
  () => groupChartDataByYValue(_points),
  [_points]
);

const longestBars = React.useMemo(
  () => computeLongestBars(groupedData, X_ORIGIN),
  [groupedData, X_ORIGIN]
);

File: packages/charts/react-charts/library/src/components/HorizontalBarChartWithAxis/HorizontalBarChartWithAxis.tsx [hunk 3]
```diff
@@ -285,21 +323,46 @@ export const HorizontalBarChartWithAxis: React.FunctionComponent<HorizontalBarCh
     containerWidth: number,
     xElement: SVGElement,
     yElement: SVGElement,
+    singleBarData: HorizontalBarChartWithAxisDataPoint[],
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    xBarScale: any,
```
Comment (srmukher) [inline packages/charts/react-charts/library/src/components/HorizontalBarChartWithAxis/HorizontalBarChartWithAxis.tsx:328]: can 'any' be replaced by its type?

File: packages/charts/react-charts/library/src/components/HorizontalBarChartWithAxis/HorizontalBarChartWithAxis.tsx [hunk 4]
```diff
@@ -494,18 +616,22 @@ export const HorizontalBarChartWithAxis: React.FunctionComponent<HorizontalBarCh
...
     const { useSingleColor } = props;
     const actions: Legend[] = [];
+    const mapLegendToColor: Record<string, string> = {};
 
     data.forEach((point: HorizontalBarChartWithAxisDataPoint, _index: number) => {
       // eslint-disable-next-line @typescript-eslint/no-shadow
       const color: string = useSingleColor ? (props.colors ? _createColors()(1) : getNextColor(1, 0)) : point.color!;
 
+      mapLegendToColor[point.legend!] = color;
+    });
```
Comment (srmukher) [inline packages/charts/react-charts/library/src/components/HorizontalBarChartWithAxis/HorizontalBarChartWithAxis.tsx:626]: can this not override a legend color if it has already been processed earlier?

File: packages/charts/react-charts/library/src/components/HorizontalBarChartWithAxis/HorizontalBarChartWithAxis.test.tsx [hunk 5]
```diff
@@ -185,10 +185,10 @@ describe('Horizontal bar chart with axis - Subcomponent bar', () => {
...
-      expect(bars[0].getAttribute('fill')).toEqual('#00bcf2');
-      expect(bars[1].getAttribute('fill')).toEqual('#00188f');
-      expect(bars[2].getAttribute('fill')).toEqual('#002050');
-      expect(bars[3].getAttribute('fill')).toEqual('#0078d4');
+      expect(bars[0].getAttribute('fill')).toEqual('#0078d4');
+      expect(bars[1].getAttribute('fill')).toEqual('#002050');
+      expect(bars[2].getAttribute('fill')).toEqual('#00188f');
+      expect(bars[3].getAttribute('fill')).toEqual('#00bcf2');
     },
   );
```
Comment (srmukher) [inline packages/charts/react-charts/library/src/components/HorizontalBarChartWithAxis/HorizontalBarChartWithAxis.test.tsx:270]: you can add tests for rendering stacked bars with positive, negative and mixed values

File: packages/charts/react-charts/library/src/components/HorizontalBarChartWithAxis/HorizontalBarChartWithAxis.tsx (hunk 2 above)
Comment (AtishayMsft) [reply packages/charts/react-charts/library/src/components/HorizontalBarChartWithAxis/HorizontalBarChartWithAxis.tsx:177]: expensive computations should be memoized only if they are to be used multiple times. Memoization has its own associated cost. See

image

File: packages/charts/react-charts/library/src/components/HorizontalBarChartWithAxis/HorizontalBarChartWithAxis.tsx (hunk 4 above)
Comment (Anush2303) [reply packages/charts/react-charts/library/src/components/HorizontalBarChartWithAxis/HorizontalBarChartWithAxis.tsx:626]: Yes, it will but for that case where there are multiple points with same legend and different colors. Firstly, I doubt if such a case will exist. Second, in that case we will have to keep either first encountered color or last.

File: packages/charts/react-charts/library/src/components/HorizontalBarChartWithAxis/HorizontalBarChartWithAxis.test.tsx (hunk 5 above)
Comment (Anush2303) [reply packages/charts/react-charts/library/src/components/HorizontalBarChartWithAxis/HorizontalBarChartWithAxis.test.tsx:270]: okay

File: packages/charts/react-charts/library/src/components/HorizontalBarChartWithAxis/HorizontalBarChartWithAxis.tsx (hunk 3 above)
Comment (Anush2303) [reply packages/charts/react-charts/library/src/components/HorizontalBarChartWithAxis/HorizontalBarChartWithAxis.tsx:328]: We can replace, but currently everywhere we are typecasting scales to type 'any' only to avoid ts errors

**Review by AtishayMsft — APPROVED** [review]
:shipit:
